File: backend/app/api/v1/businesses.py
import math
import uuid
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.models.user import User
from app.schemas.business import (
    BusinessCreate,
    BusinessDuplicateCheckResponse,
    BusinessResponse,
    BusinessUpdate,
)
from app.schemas.common import DataResponse, PaginatedResponse, PaginationMeta
from app.services.business import BusinessService

router = APIRouter(prefix="/businesses", tags=["Business CRM"])


@router.post("", response_model=DataResponse[BusinessResponse], status_code=201)
def create_business(
    data: BusinessCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    business = BusinessService.create_business(db, data, user_id=current_user.id)
    quality = BusinessService.calculate_data_quality(business)
    resp = BusinessResponse.model_validate(business)
    resp.data_quality = quality
    return {"data": resp}


@router.get("", response_model=PaginatedResponse[BusinessResponse])
def list_businesses(
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=25, ge=1, le=100),
    search: Optional[str] = Query(default=None),
    status: Optional[str] = Query(default=None),
    industry: Optional[str] = Query(default=None),
    business_type: Optional[str] = Query(default=None),
    city: Optional[str] = Query(default=None),
    country: Optional[str] = Query(default=None),
    source: Optional[str] = Query(default=None),
    sort: str = Query(default="created_at"),
    order: str = Query(default="desc"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    items, total = BusinessService.list_businesses(
        db=db,
        page=page,
        page_size=page_size,
        search=search,
        status=status,
        industry=industry,
        business_type=business_type,
        city=city,
        country=country,
        source=source,
        sort=sort,
        order=order,
    )

    business_responses = []
    for item in items:
        resp = BusinessResponse.model_validate(item)
        resp.data_quality = BusinessService.calculate_data_quality(item)
        business_responses.append(resp)

    total_pages = math.ceil(total / page_size) if page_size > 0 else 0

    return {
        "data": business_responses,
        "pagination": PaginationMeta(
            page=page,
            page_size=page_size,
            total=total,
            total_pages=total_pages,
        ),
    }


@router.post("/check-duplicate", response_model=DataResponse[BusinessDuplicateCheckResponse])
def check_duplicate_business(
    name: str = Query(..., min_length=1),
    phone: Optional[str] = Query(default=None),
    email: Optional[str] = Query(default=None),
    website_url: Optional[str] = Query(default=None),
    city: Optional[str] = Query(default=None),
    source: Optional[str] = Query(default=None),
    external_id: Optional[str] = Query(default=None),
    exclude_id: Optional[uuid.UUID] = Query(default=None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    dup_result = BusinessService.check_duplicates(
        db=db,
        name=name,
        phone=phone,
        email=email,
        website_url=website_url,
        city=city,
        source=source,
        external_id=external_id,
        exclude_id=exclude_id,
    )
    return {"data": dup_result}


@router.get("/{business_id}", response_model=DataResponse[BusinessResponse])
def get_business(
    business_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    business = BusinessService.get_business(db, business_id)
    quality = BusinessService.calculate_data_quality(business)
    resp = BusinessResponse.model_validate(business)
    resp.data_quality = quality
    return {"data": resp}


@router.patch("/{business_id}", response_model=DataResponse[BusinessResponse])
def update_business(
    business_id: uuid.UUID,
    data: BusinessUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    business = BusinessService.update_business(db, business_id, data)
    quality = BusinessService.calculate_data_quality(business)
    resp = BusinessResponse.model_validate(business)
    resp.data_quality = quality
    return {"data": resp}


@router.delete("/{business_id}", response_model=DataResponse[BusinessResponse])
def archive_business_by_delete(
    business_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    business = BusinessService.archive_business(db, business_id)
    quality = BusinessService.calculate_data_quality(business)
    resp = BusinessResponse.model_validate(business)
    resp.data_quality = quality
    return {"data": resp}


@router.post("/{business_id}/archive", response_model=DataResponse[BusinessResponse])
def archive_business(
    business_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    business = BusinessService.archive_business(db, business_id)
    quality = BusinessService.calculate_data_quality(business)
    resp = BusinessResponse.model_validate(business)
    resp.data_quality = quality
    return {"data": resp}


@router.post("/{business_id}/restore", response_model=DataResponse[BusinessResponse])
def restore_business(
    business_id: uuid.UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    business = BusinessService.restore_business(db, business_id)
    quality = BusinessService.calculate_data_quality(business)
    resp = BusinessResponse.model_validate(business)
    resp.data_quality = quality
    return {"data": resp}

File: frontend/lib/api/businesses.ts
import { api } from "./client";

export interface BusinessDataQuality {
  score: number;
  missing_fields: string[];
  warnings: string[];
}

export interface Business {
  id: string;
  name: string;
  normalized_name: string;
  legal_name?: string | null;
  description?: string | null;
  business_type: string;
  industry: string;
  category?: string | null;
  subcategory?: string | null;
  phone?: string | null;
  normalized_phone?: string | null;
  email?: string | null;
  normalized_email?: string | null;
  website_url?: string | null;
  normalized_website?: string | null;
  address?: string | null;
  city?: string | null;
  state?: string | null;
  country?: string | null;
  postal_code?: string | null;
  latitude?: number | null;
  longitude?: number | null;
  timezone?: string | null;
  status: "ACTIVE" | "INACTIVE" | "ARCHIVED";
  source: string;
  source_url?: string | null;
  external_id?: string | null;
  created_by_user_id?: string | null;
  archived_at?: string | null;
  created_at: string;
  updated_at: string;
  data_quality?: BusinessDataQuality | null;
}

export interface BusinessCreateInput {
  name: string;
  legal_name?: string;
  description?: string;
  business_type?: string;
  industry?: string;
  category?: string;
  subcategory?: string;
  phone?: string;
  email?: string;
  website_url?: string;
  address?: string;
  city?: string;
  state?: string;
  country?: string;
  postal_code?: string;
  source?: string;
  source_url?: string;
  external_id?: string;
}

export interface BusinessUpdateInput extends Partial<BusinessCreateInput> {
  status?: "ACTIVE" | "INACTIVE" | "ARCHIVED";
}

export interface BusinessDuplicateCheckResponse {
  possible_duplicate: boolean;
  confidence: number;
  signals: string[];
  matches: Array<{
    business_id: string;
    name: string;
    confidence: number;
    signals: string[];
  }>;
}

export interface BusinessListParams {
  page?: number;
  page_size?: number;
  search?: string;
  status?: string;
  industry?: string;
  business_type?: string;
  city?: string;
  country?: string;
  source?: string;
  sort?: string;
  order?: "asc" | "desc";
}

export interface PaginatedBusinessesResponse {
  data: Business[];
  pagination: {
    page: number;
    page_size: number;
    total: number;
    total_pages: number;
  };
}

export async function listBusinesses(params: BusinessListParams = {}): Promise<PaginatedBusinessesResponse> {
  const query = new URLSearchParams();
  if (params.page) query.append("page", params.page.toString());
  if (params.page_size) query.append("page_size", params.page_size.toString());
  if (params.search) query.append("search", params.search);
  if (params.status) query.append("status", params.status);
  if (params.industry) query.append("industry", params.industry);
  if (params.business_type) query.append("business_type", params.business_type);
  if (params.city) query.append("city", params.city);
  if (params.country) query.append("country", params.country);
  if (params.source) query.append("source", params.source);
  if (params.sort) query.append("sort", params.sort);
  if (params.order) query.append("order", params.order);

  const queryString = query.toString() ? `?${query.toString()}` : "";
  return api<PaginatedBusinessesResponse>(`/businesses${queryString}`);
}

export async function getBusiness(id: string): Promise<{ data: Business }> {
  return api<{ data: Business }>(`/businesses/${id}`);
}

export async function createBusiness(input: BusinessCreateInput): Promise<{ data: Business }> {
  return api<{ data: Business }>("/businesses", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function updateBusiness(id: string, input: BusinessUpdateInput): Promise<{ data: Business }> {
  return api<{ data: Business }>(`/businesses/${id}`, {
    method: "PATCH",
    body: JSON.stringify(input),
  });
}

export async function archiveBusiness(id: string): Promise<{ data: Business }> {
  return api<{ data: Business }>(`/businesses/${id}/archive`, {
    method: "POST",
  });
}

export async function restoreBusiness(id: string): Promise<{ data: Business }> {
  return api<{ data: Business }>(`/businesses/${id}/restore`, {
    method: "POST",
  });
}

export async function checkDuplicateBusiness(name: string, extra: Record<string, string> = {}): Promise<{ data: BusinessDuplicateCheckResponse }> {
  const query = new URLSearchParams({ name, ...extra });
  return api<{ data: BusinessDuplicateCheckResponse }>(`/businesses/check-duplicate?${query.toString()}`);
}
